File: threads/src/main/java/im/threads/internal/activities/FilesActivity.java
package im.threads.internal.activities;

import android.content.Context;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.TypedValue;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import im.threads.ChatStyle;
import im.threads.R;
import im.threads.internal.Config;
import im.threads.internal.adapters.FilesAndMediaAdapter;
import im.threads.internal.broadcastReceivers.ProgressReceiver;
import im.threads.internal.controllers.FilesAndMediaController;
import im.threads.internal.model.FileDescription;
import im.threads.internal.utils.Keyboard;

public final class FilesActivity extends BaseActivity implements FilesAndMediaAdapter.OnFileClick, ProgressReceiver.Callback {
    private static final String TAG = "FilesActivity ";
    private FilesAndMediaController mFilesAndMediaController;
    private RecyclerView mRecyclerView;
    private View mEmptyListLayout;
    private View mSearchImageView;
    private EditText mSearchEditText;
    private Toolbar mToolbar;
    private FilesAndMediaAdapter mFilesAndMediaAdapter;

    public static Intent getStartIntent(@NonNull Context context) {
        return new Intent(context, FilesActivity.class);
    }

    public void onFileReceive(List<FileDescription> descriptions) {
        if (descriptions != null && descriptions.size() > 0) {
            mSearchImageView.setVisibility(View.VISIBLE);
            mEmptyListLayout.setVisibility(View.GONE);
            mRecyclerView.setVisibility(View.VISIBLE);
            mFilesAndMediaAdapter = new FilesAndMediaAdapter(descriptions, this, this);
            mRecyclerView.setAdapter(mFilesAndMediaAdapter);
        }
    }

    @Override
    public void onFileClick(FileDescription fileDescription) {
        mFilesAndMediaController.onFileClick(fileDescription);
    }

    @Override
    public void onDownloadFileClick(FileDescription fileDescription) {
        mFilesAndMediaController.onDownloadFileClick(fileDescription);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ChatStyle style = Config.instance.getChatStyle();
        Window window = getWindow();
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.setStatusBarColor(getResources().getColor(style.mediaAndFilesStatusBarColorResId));
        if (getResources().getBoolean(style.mediaAndFilesWindowLightStatusBarResId)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
            }
        }
        setContentView(R.layout.activity_files_and_media);
        mToolbar = findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
        mToolbar.setNavigationOnClickListener(v -> onBackPressed());
        if (getSupportFragmentManager().findFragmentByTag(TAG) == null) {
            mFilesAndMediaController = FilesAndMediaController.getInstance();
            getSupportFragmentManager().beginTransaction().add(mFilesAndMediaController, TAG).commit();
        } else {
            mFilesAndMediaController = (FilesAndMediaController) getSupportFragmentManager().findFragmentByTag(TAG);
        }
        mFilesAndMediaController.bindActivity(this);
        mFilesAndMediaController.getFilesAsync();
        mEmptyListLayout = findViewById(R.id.empty_list_layout);
        mRecyclerView = findViewById(R.id.recycler);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        mSearchEditText = findViewById(R.id.search_edit_text);
        mSearchImageView = findViewById(R.id.search);
        mSearchImageView.setOnClickListener(v -> {
            if (mSearchEditText.getVisibility() == View.VISIBLE) {
                mSearchEditText.setText("");
                mSearchEditText.setVisibility(View.GONE);
                mToolbar.setTitle(getString(R.string.threads_files_and_media));
                if (null != mFilesAndMediaAdapter) {
                    mFilesAndMediaAdapter.undoClear();
                }
            } else {
                mSearchEditText.setVisibility(View.VISIBLE);
                mSearchEditText.requestFocus();
                mToolbar.setTitle("");
                if (null != mFilesAndMediaAdapter) {
                    mFilesAndMediaAdapter.backupAndClear();
                }
                mSearchEditText.setText("");
                Keyboard.show(this, mSearchEditText, 100);
            }
        });
        mSearchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                String str;
                if (s == null) {
                    str = "";
                } else {
                    str = s.toString();
                }
                search(str);
            }
        });
        mSearchEditText.setOnEditorActionListener((v, actionId, event) -> {
            if ((mSearchEditText.getVisibility() == View.VISIBLE) && actionId == EditorInfo.IME_ACTION_SEARCH) {
                search(v.getText().toString());
                return true;
            } else {
                return false;
            }
        });
        setActivityStyle(Config.instance.getChatStyle());
    }

    @Override
    public void onBackPressed() {
        if (mSearchEditText.getVisibility() == View.VISIBLE) {
            mSearchEditText.setText("");
            mSearchEditText.setVisibility(View.GONE);
            mToolbar.setTitle(getString(R.string.threads_files_and_media));
            if (null != mFilesAndMediaAdapter) {
                mFilesAndMediaAdapter.undoClear();
            }
        } else {
            super.onBackPressed();
        }
    }

    private void search(String searchString) {
        if (mFilesAndMediaAdapter != null) {
            mFilesAndMediaAdapter.filter(searchString);
        }
    }

    private void setActivityStyle(@NonNull ChatStyle style) {
        findViewById(R.id.activity_root).setBackgroundColor(ContextCompat.getColor(this, style.mediaAndFilesScreenBackgroundColor));
        mToolbar.setBackgroundColor(ContextCompat.getColor(this, style.mediaAndFilesToolbarColorResId));
        ((ImageButton) findViewById(R.id.search)).setColorFilter(ContextCompat.getColor(this, style.mediaAndFilesToolbarTextColorResId), PorterDuff.Mode.SRC_ATOP);
        mSearchEditText.setTextColor(getColorInt(style.mediaAndFilesToolbarTextColorResId));
        final Drawable navigationIcon = mToolbar.getNavigationIcon();
        navigationIcon.mutate().setColorFilter(ContextCompat.getColor(this, style.mediaAndFilesToolbarTextColorResId), PorterDuff.Mode.SRC_ATOP);
        mToolbar.setNavigationIcon(navigationIcon);
        mSearchEditText.setHintTextColor(getColorInt(style.mediaAndFilesToolbarHintTextColor));

        setUpTextViewStyle(R.id.empty_list_header,
                style.emptyMediaAndFilesHeaderText,
                style.emptyMediaAndFilesHeaderTextSize,
                style.emptyMediaAndFilesHeaderTextColor,
                style.emptyMediaAndFilesHeaderFontPath);
        setUpTextViewStyle(R.id.empty_list_description,
                style.emptyMediaAndFilesDescriptionText,
                style.emptyMediaAndFilesDescriptionTextSize,
                style.emptyMediaAndFilesDescriptionTextColor,
                style.emptyMediaAndFilesDescriptionFontPath);
    }

    private void setUpTextViewStyle(@IdRes int textViewResId,
                                    @StringRes int textResId,
                                    @DimenRes int sizeResId,
                                    @ColorRes int colorResId,
                                    @Nullable String fontPath) {
        TextView textView = findViewById(textViewResId);
        textView.setText(textResId);
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimension(sizeResId));
        textView.setTextColor(getColorInt(colorResId));
        if (!TextUtils.isEmpty(fontPath)) {
            textView.setTypeface(Typeface.createFromAsset(getAssets(), fontPath));
        }
    }

    @Override
    public void updateProgress(FileDescription fileDescription) {
        mFilesAndMediaAdapter.updateProgress(fileDescription);
    }

    @Override
    public void onDownloadError(FileDescription fileDescription, Throwable t) {
        mFilesAndMediaAdapter.onDownloadError(fileDescription);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mFilesAndMediaController.unbindActivity();
    }
}

File: threads/src/main/java/im/threads/internal/activities/filesActivity/FilesActivity.kt
package im.threads.internal.activities.filesActivity

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.PorterDuff
import android.graphics.Typeface
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.TypedValue
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import im.threads.ChatStyle
import im.threads.R
import im.threads.business.models.FileDescription
import im.threads.business.secureDatabase.DatabaseHolder
import im.threads.databinding.ActivityFilesAndMediaBinding
import im.threads.internal.Config
import im.threads.internal.activities.BaseActivity
import im.threads.internal.adapters.filesAndMedia.FilesAndMediaAdapter
import im.threads.internal.adapters.filesAndMedia.FilesAndMediaAdapter.OnFileClick
import im.threads.internal.utils.Keyboard
import im.threads.internal.utils.setColorFilter

/**
 * Показывает список файлов, которые присутствовали в диалоге с оператором с обоих сторон
 */
internal class FilesActivity : BaseActivity(), OnFileClick {
    private val binding: ActivityFilesAndMediaBinding by lazy {
        ActivityFilesAndMediaBinding.inflate(layoutInflater)
    }
    private val filesViewModel: FilesViewModel by viewModels {
        FilesViewModel.Factory(Config.instance.context, DatabaseHolder.getInstance())
    }
    private val style = Config.instance.chatStyle
    private var filesAndMediaAdapter: FilesAndMediaAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        hideSystemUI()
        setStatusBarColor()
        setContentView(binding.root)
        initToolbar()
        setOnSearchClickAction()
        setOnSearchTextChanged()
        setActivityStyle(Config.instance.chatStyle)
        subscribeForNewIntents()
        subscribeForDownloadProgress()
        requestFiles()
    }

    override fun onBackPressed() {
        if (binding.searchEditText.visibility == View.VISIBLE) {
            binding.searchEditText.setText("")
            binding.searchEditText.visibility = View.GONE
            binding.toolbar.title = getString(R.string.threads_files_and_media)
            filesAndMediaAdapter?.undoClear()
        } else {
            super.onBackPressed()
        }
    }

    override fun onFileClick(fileDescription: FileDescription?) {
        filesViewModel.onFileClick(fileDescription)
    }

    override fun onDownloadFileClick(fileDescription: FileDescription?) {
        filesViewModel.onDownloadFileClick(fileDescription)
    }

    private fun onFileReceive(descriptions: List<FileDescription?>?) = with(binding) {
        if (descriptions != null && descriptions.isNotEmpty()) {
            searchButton.visibility = View.VISIBLE
            emptyListLayout.visibility = View.GONE
            filesRecycler.visibility = View.VISIBLE
            filesAndMediaAdapter = FilesAndMediaAdapter(
                descriptions,
                this@FilesActivity,
                this@FilesActivity
            )
            filesRecycler.adapter = filesAndMediaAdapter
        }
    }

    @Suppress("DEPRECATION")
    private fun hideSystemUI() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.setDecorFitsSystemWindows(true)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        }
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
    }

    private fun setStatusBarColor() {
        val statusBarColor = ContextCompat.getColor(baseContext, style.mediaAndFilesStatusBarColorResId)
        val isStatusBarLight = resources.getBoolean(style.mediaAndFilesWindowLightStatusBarResId)

        super.setStatusBarColor(isStatusBarLight, statusBarColor)
    }

    private fun initToolbar() = with(binding) {
        setSupportActionBar(toolbar)
        toolbar.setNavigationOnClickListener { onBackPressed() }
        val isShadowVisible = resources.getBoolean(style.isChatTitleShadowVisible)
        toolbarShadow.visibility = if (isShadowVisible) View.VISIBLE else View.INVISIBLE
        if (!isShadowVisible) {
            val noElevation = 0f
            toolbar.elevation = noElevation
        }
    }

    private fun setOnSearchClickAction() = with(binding) {
        searchButton.setOnClickListener {
            if (searchEditText.visibility == View.VISIBLE) {
                searchEditText.setText("")
                searchEditText.visibility = View.GONE
                toolbar.title = getString(R.string.threads_files_and_media)
                filesAndMediaAdapter?.undoClear()
            } else {
                searchEditText.visibility = View.VISIBLE
                searchEditText.requestFocus()
                toolbar.title = ""
                filesAndMediaAdapter?.backupAndClear()
                searchEditText.setText("")
                Keyboard.show(baseContext, searchEditText, 100)
            }
        }
    }

    private fun setOnSearchTextChanged() = with(binding) {
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(text: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(text: Editable) {
                search(text.toString())
            }
        })
        searchEditText.setOnEditorActionListener { textView: TextView, actionId: Int, _: KeyEvent? ->
            if (searchEditText.visibility == View.VISIBLE && actionId == EditorInfo.IME_ACTION_SEARCH) {
                search(textView.text.toString())
                return@setOnEditorActionListener true
            } else {
                return@setOnEditorActionListener false
            }
        }
    }

    private fun requestFiles() {
        filesViewModel.getFilesAsync()
    }

    private fun search(searchString: String) {
        filesAndMediaAdapter?.filter(searchString)
    }

    private fun setActivityStyle(style: ChatStyle) = with(binding) {
        root.setBackgroundColor(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesScreenBackgroundColor
            )
        )
        toolbar.setBackgroundColor(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesToolbarColorResId
            )
        )
        searchButton.setColorFilter(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesToolbarTextColorResId
            ),
            PorterDuff.Mode.SRC_ATOP
        )
        searchEditText.setTextColor(getColorInt(style.mediaAndFilesToolbarTextColorResId))
        toolbar.navigationIcon?.mutate()?.setColorFilter(
            ContextCompat.getColor(baseContext, style.mediaAndFilesToolbarTextColorResId)
        )
        searchEditText.setHintTextColor(getColorInt(style.mediaAndFilesToolbarHintTextColor))
        setUpTextViewStyle(
            emptyListHeader,
            style.emptyMediaAndFilesHeaderText,
            style.emptyMediaAndFilesHeaderTextSize,
            style.emptyMediaAndFilesHeaderTextColor,
            style.emptyMediaAndFilesHeaderFontPath
        )
        setUpTextViewStyle(
            emptyListDescription,
            style.emptyMediaAndFilesDescriptionText,
            style.emptyMediaAndFilesDescriptionTextSize,
            style.emptyMediaAndFilesDescriptionTextColor,
            style.emptyMediaAndFilesDescriptionFontPath
        )
    }

    private fun subscribeForNewIntents() {
        filesViewModel.intentLiveData.observe(this) { startIntent(it) }
    }

    private fun subscribeForDownloadProgress() {
        filesViewModel.filesFlowLiveData.observe(this) { onDownloadProgress(it) }
    }

    private fun setUpTextViewStyle(
        textView: TextView,
        @StringRes textResId: Int,
        @DimenRes sizeResId: Int,
        @ColorRes colorResId: Int,
        fontPath: String?
    ) = with(binding) {
        textView.setText(textResId)
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimension(sizeResId))
        textView.setTextColor(getColorInt(colorResId))
        if (!TextUtils.isEmpty(fontPath)) {
            textView.typeface = Typeface.createFromAsset(assets, fontPath)
        }
    }

    private fun startIntent(intent: Intent) {
        try {
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            Toast.makeText(
                baseContext,
                getString(R.string.threads_file_not_supported),
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private fun onDownloadProgress(filesFlow: FilesFlow) {
        when (filesFlow) {
            is FilesFlow.UpdatedProgress -> {
                filesAndMediaAdapter?.updateProgress(filesFlow.fileDescription)
            }
            is FilesFlow.DownloadError -> {
                filesAndMediaAdapter?.onDownloadError(filesFlow.fileDescription)
            }
            is FilesFlow.FilesReceived -> {
                onFileReceive(filesFlow.files)
            }
        }
    }

    companion object {
        /**
         * Запускает FilesActivity
         * @param activity активити, из которой будет произведен запуск FilesActivity.
         */
        @JvmStatic
        fun startActivity(activity: Activity?) {
            activity?.startActivity(Intent(activity, FilesActivity::class.java))
        }
    }
}
